// @tsed/schema/lib/domain/DecoratorContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecoratorContext = void 0;
const core_1 = require("@tsed/core");
const change_case_1 = require("change-case");
/**
 * @ignore
 */
class DecoratorContext extends Map {
    constructor() {
        super(...arguments);
        this.actions = [];
    }
    addAction(cb) {
        this.actions.push(cb);
        return this;
    }
    build() {
        const decorator = (...args) => this.onInit(args, decorator);
        const wrap = (cb) => {
            return (...args) => {
                cb(...args);
                return decorator;
            };
        };
        const wrapKey = (key) => {
            return wrap((...values) => {
                key in this ? this[key](...values) : this.set(key, values[0]);
            });
        };
        this.methods.forEach((name) => {
            decorator[change_case_1.pascalCase(name)] = wrapKey(name);
        });
        return decorator;
    }
    toObject() {
        return [...this.entries()].reduce((obj, [key, value]) => {
            return {
                ...obj,
                [key]: value
            };
        }, {});
    }
    getMergedKey(key, defaultValue) {
        let value = this.get(key) || defaultValue;
        if (!value) {
            return;
        }
        if (this.decoratorType === core_1.DecoratorTypes.CLASS) {
            value = {
                ...value,
                ...(defaultValue || {})
            };
        }
        return {
            ...(defaultValue || {}),
            ...value
        };
    }
    runActions() {
        this.actions.forEach((action) => {
            action(this);
        });
        return this;
    }
}
exports.DecoratorContext = DecoratorContext;
