// @tsed/common/lib/platform/domain/PlatformRouteDetails.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformRouteDetails = void 0;
const ParamTypes_1 = require("../../mvc/models/ParamTypes");
class PlatformRouteDetails {
    constructor({ endpoint, method, url }) {
        this.endpoint = endpoint;
        this.method = method;
        this.url = url;
        this.rawBody = !!endpoint.params.find((param) => param.paramType === ParamTypes_1.ParamTypes.RAW_BODY);
    }
    get name() {
        return `${this.endpoint.targetName}.${this.methodClassName}()`;
    }
    get className() {
        return this.endpoint.targetName;
    }
    get methodClassName() {
        return String(this.endpoint.propertyKey);
    }
    get parameters() {
        return this.endpoint.params;
    }
    toJSON() {
        return {
            method: this.method,
            name: this.name,
            url: this.url,
            className: this.className,
            methodClassName: this.methodClassName,
            parameters: this.parameters,
            rawBody: this.rawBody
        };
    }
}
exports.PlatformRouteDetails = PlatformRouteDetails;
